C7GameData/Tile.cs

```diff
@@ -55,12 +58,148 @@ public class Tile
 
 		public TileOverlays overlays = new TileOverlays();
 
+
+		// -1 = not initialized. 0 = false, 1 = true
+		public int isLake = -1;
+
 		public Tile()
 		{
 			unitsOnTile = new List<MapUnit>();
 			Resource = Resource.NONE;
 		}
 
+		public bool IsLake()
+		{
+			// If the isLake value has been initialized, we should return that
+			if(isLake != -1)
+			{
+				return (isLake == 1);
+			}
+
+			if(!IsWater())
+			{
+				isLake = 0;
+				return false;
+			}
+
+			int contiguousWater = 0;
+			List<Tile> contiguous = new List<Tile>();
+			List<Tile> searchedWater = new List<Tile>();
+			contiguous.Add(this);
+
+			// Default is that this *is* a lake because other conditions will assure we only get water tiles. If the lake is under size 21, we want to default to it being a lake
+			bool lake = true;
+
+			int i = 0;
+			while(i < contiguous.Count())
+			{
+				Tile currentTile = contiguous[i];
+
+				// Skip land tiles and remove from the stack. This should be checked before they're added but just in case
+				if(!currentTile.IsWater())
+				{
+					currentTile.isLake = 0;
+					contiguous.Remove(currentTile);
+					continue;
+				}
+				contiguousWater += 1;
+				if(contiguousWater > 20)
+				{
+					lake = false;
+					break;
+				}
+
+				// If a contiguous tile is a lake, this (and all contiguous water) is too. This works because contiguousness is transitive. If it is *not* a lake (and is a water tile), then this isn't either.
+				if(currentTile.isLake == 1)
+				{
+					lake = true;
+					break;
+				}
+				if(currentTile.isLake == 0)
+				{
+					lake = false;
+					break;
+				}
+				// If we haven't found out one way or the other, add all of this tile's water neighbors to the stack and then remove this tile
+				foreach((TileDirection dir, Tile neighbor) in currentTile.neighbors)
+				{
+					if(!neighbor.IsWater())
+					{
+						continue;
+					}
+					else if(contiguous.Contains(neighbor) || searchedWater.Contains(neighbor))
+					{
+						continue;
+					}
+					else
+					{
+						// Add neighbor to tiles we should search as long as there's no isthmus blocking
+						if(currentTile.isthmusAt(dir))
+						{
+							continue;
+						}
+						else
+						{
+							contiguous.Add(neighbor);
+						}
+					}
+				}
+				i ++;
+			}
+			if(lake)
+			{
+				// Update all the contiguous water tiles we've looked at. It won't be the whole body of water but it'll save us some repeats.
+				foreach(Tile currentTile in contiguous)
+				{
+					if(currentTile.IsWater())
+					{
+						currentTile.isLake = 1;
+					}
+				}
+				return true;
+			}
+			else
+			{
```

Comment:
I'm pretty sure that this block needs to check the value of the `contiguousWater` variable (or what it eventually is named, see above).  In the case that there are fewer than LAKE_THRESHOLD tiles examined, it's a lake, right?  But if we don't already know that one of the other tiles is a lake, it won't pass the `if` check on line 149.

Maybe the change should be on line 149, checking `if (lake || contiguousWater < LAKE_THRESHOLD)`?

This method is also ripe for a test in C7GameDataTests.  Not saying it needs to be added now, and we might not have enough convenience methods to make it easy to set up the needed data at this point, but the logic is just complex enough that it's a good example of where a test would be high-value.